Feature: ProxySQL local DNS Cache

rahim-kanji opened the pull request:
Local DNS Cache for ProxySQL

* Supports IPv4 and IPv6
* Multiple IP support to load balance connections (Round-Robin Scheduling).
* Scalable DNS resolver (can scale upto 32 Threads).
* Prometheus metrics:  
* *proxysql_mysql_monitor_dns_cache_lookup_success 
* *proxysql_mysql_monitor_dns_cache_queried
* *proxysql_mysql_monitor_dns_cache_record_updated
* ProxySQL variable to configure DNS Cache:
* *mysql-monitor_local_dns_cache_ttl (0 = DNS Cache disabled)
* *mysql-monitor_local_dns_cache_refresh_interval (0 = DNS Cache disabled)
* *mysql-monitor_local_dns_resolver_queue_maxsize

rahim-kanji committed:
Feature: ProxySQL local DNS Cache

rahim-kanji committed:
Optimization: moving DNS_Record object instead of copying

rahim-kanji committed:
Added DNS Cache enable/disable flag handling.

rahim-kanji committed:
Merge remote-tracking branch 'Master/v2.x' into v2.x

rahim-kanji committed:
No need to do lookup if IP was provided

rahim-kanji committed:
Heap allocated host pointer for async mysql_real_connect_start function
Few fixes

rahim-kanji committed:
* Added hostname and ip information in mysql_connection
* Added multiple IP support with load balancing (Round Robin Scheduling)
* Added DNS resolver request max queue size
* Added AI_ADDRCONFIG flag
* Exception handling

rahim-kanji committed:
Removed hard coded env variables

rahim-kanji committed:
Replaced SELECT 1 with DO 1 to route traffic to hostgroup 0 (writer)

mirostauder commented:
retest this please

mirostauder commented:
retest this please

rahim-kanji committed:
TAP Test: Adding dummy server with host group id 999 and setting all users default host group to 999

JavierJF commented:
retest this please

rahim-kanji committed:
Merge remote-tracking branch 'Master/v2.x' into v2.x_LocalDNSCache

JavierJF commented:
retest this please

JavierJF commented:
retest this please

rahim-kanji committed:
Added Debug logs
Few optimizations

rahim-kanji commented:
retest this please

rahim-kanji commented:
retest this please

rahim-kanji commented:
retest this please

renecannao commented on lib/MySQL_Session.cpp:
```diff
@@ -218,7 +218,7 @@ void * kill_query_thread(void *arg) {
 				break;
 		}
-		ret=mysql_real_connect(mysql,ka->hostname,ka->username,ka->password,NULL,ka->port,NULL,0);
+		ret=mysql_real_connect(mysql, MySQL_Monitor::dns_lookup(ka->hostname).c_str(), ka->username, ka->password, NULL, ka->port, NULL, 0);
```
Maybe instead of the `hostname` here we should use the IP that is already stored in `connected_host_details.ip`

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -100,7 +100,7 @@ void * ProxySQL_Cluster_Monitor_thread(void *args) {
 			{ unsigned char val = 1; mysql_options(conn, MYSQL_OPT_SSL_ENFORCE, &val); }
 			//rc_conn = mysql_real_connect(conn, node->hostname, username, password, NULL, node->port, NULL, CLIENT_COMPRESS); // FIXME: add optional support for compression
-			rc_conn = mysql_real_connect(conn, node->hostname, username, password, NULL, node->port, NULL, 0);
+			rc_conn = mysql_real_connect(conn, MySQL_Monitor::dns_lookup(node->hostname).c_str(), username, password, NULL, node->port, NULL, 0);
```
The approach here is correct, but I think we now have the opportunity to improve cluster even further.
If more then *one* IP is present for a given ProxySQL node, we should generate a warning in the error log. Perhaps even an error!
Ideally we should also prevent to establish a connection to that IP , but this is perhaps too much.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -881,7 +881,7 @@ void ProxySQL_Cluster::pull_mysql_query_rules_from_peer(const string& expected_c
 			{ unsigned char val = 1; mysql_options(conn, MYSQL_OPT_SSL_ENFORCE, &val); }
 			proxy_info("Cluster: Fetching MySQL Query Rules from peer %s:%d started. Expected checksum: %s\n", hostname, port, expected_checksum.c_str());
-			rc_conn = mysql_real_connect(conn, hostname, username, password, NULL, port, NULL, 0);
+			rc_conn = mysql_real_connect(conn, MySQL_Monitor::dns_lookup(hostname).c_str(), username, password, NULL, port, NULL, 0);
```
We should be able to connect to the IP directly without the need for dns lookup.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -1147,7 +1147,7 @@ void ProxySQL_Cluster::pull_mysql_users_from_peer(const string& expected_checksu
 			proxy_info("Cluster: Fetching MySQL Users from peer %s:%d started. Expected checksum: %s\n", hostname, port, expected_checksum.c_str());
 
-			rc_conn = mysql_real_connect(conn, hostname, username, password, NULL, port, NULL, 0);
+			rc_conn = mysql_real_connect(conn, MySQL_Monitor::dns_lookup(hostname).c_str(), username, password, NULL, port, NULL, 0);
```
We should be able to connect to the IP directly without the need for dns lookup.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -1461,7 +1461,7 @@ void ProxySQL_Cluster::pull_mysql_servers_from_peer(const std::string& checksum,
 			{ unsigned char val = 1; mysql_options(conn, MYSQL_OPT_SSL_ENFORCE, &val); }
 			proxy_info("Cluster: Fetching MySQL Servers from peer %s:%d started. Expected checksum %s\n", hostname, port, peer_checksum);
-			rc_conn = mysql_real_connect(conn, hostname, username, password, NULL, port, NULL, 0);
+			rc_conn = mysql_real_connect(conn, MySQL_Monitor::dns_lookup(hostname).c_str(), username, password, NULL, port, NULL, 0);
```
We should be able to connect to the IP directly without the need for dns lookup.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -1825,7 +1825,7 @@ void ProxySQL_Cluster::pull_global_variables_from_peer(const string& var_type, c
 			{ unsigned char val = 1; mysql_options(conn, MYSQL_OPT_SSL_ENFORCE, &val); }
 			proxy_info("Cluster: Fetching %s variables from peer %s:%d started\n", vars_type_str, hostname, port);
-			rc_conn = mysql_real_connect(conn, hostname, username, password, NULL, port, NULL, 0);
+			rc_conn = mysql_real_connect(conn, MySQL_Monitor::dns_lookup(hostname).c_str(), username, password, NULL, port, NULL, 0);
```
Same, we should be able to connect to the IP directly without the need for dns lookup.

renecannao commented on lib/ProxySQL_Cluster.cpp:
```diff
@@ -1974,7 +1974,7 @@ void ProxySQL_Cluster::pull_proxysql_servers_from_peer(const std::string& expect
 				hostname, port, expected_checksum.c_str()
 			);
-			rc_conn = mysql_real_connect(conn, hostname, username, password, NULL, port, NULL, 0);
+			rc_conn = mysql_real_connect(conn, MySQL_Monitor::dns_lookup(hostname).c_str(), username, password, NULL, port, NULL, 0);
```
As above, we should be able to connect to the IP directly without the need for dns lookup.

renecannao commented on lib/mysql_connection.cpp:
```diff
@@ -817,7 +823,38 @@ void MySQL_Connection::connect_start() {
+			if (connected_host_details.hostname) {
+				if (strcmp(connected_host_details.hostname, parent->address) != 0) {
+					connected_host_details.hostname = (char*)realloc(connected_host_details.hostname, strlen(parent->address) + 1);
+					strcpy(connected_host_details.hostname, parent->address);
```
I see a mix of `realloc` + `strcpy` , or `strdup` .
Personally I find more clear to always use `strdup` , and `free` when needed.

rahim-kanji committed:
* Added DNS cache monitoring for ProxySQL servers.
* Added forced triggering of DNS Cache Update on MySQL Servers and ProxySQL Servers table update.
* Direct DNS cache update via socket (connected peer ip) if record is not available in cache.
* Used strdup instead of realloc.
* DNS lookup only once for ProxySQL Cluster Nodes.

rahim-kanji commented:
retest this please

rahim-kanji committed:
Code cleanup

renecannao committed:
Replaced an error with a warning

In ProxySQL_Node_Entry::ProxySQL_Node_Entry() replaced a proxy_error with a proxy_warning, and more verbose message
